keras/engine/training.py

```diff
@@ -3130,12 +3130,39 @@ def to_yaml(self, **kwargs):
             "arbitrary code execution. Please use `model.to_json()` instead."
         )
 
-    def reset_states(self):
+    def reset_states(self, states=None):
+        """Reset the recorded states for the stateful RNN layers.
+
+        Can only be used when RNN layers are constructed with `stateful` =
+        `True`.
+        Args:
+          states: Numpy arrays that contains the value for the initial state,
+            which will be feed to cell at the first time step. When the value is
+            None, zero filled numpy array will be created based on the cell
+            state size.
+
+        Raises:
+          AttributeError: When the RNN layer is not stateful. ValueError: When
+          the batch size of the RNN layer is unknown. ValueError: When the input
+          numpy array is not compatible with the RNN
+            layer state, either size wise or dtype wise.
+        """
+        if states is not None and not isinstance(states, (list, tuple)):
+            raise ValueError(
+                "Expected states to be None, list or tuple but is "
+                f"{type(states)}"
+            )
+
+        start = 0
         for layer in self.layers:
             if hasattr(layer, "reset_states") and getattr(
                 layer, "stateful", False
             ):
-                layer.reset_states()
+                end = start + len(layer.states)
+                layer.reset_states(
+                    None if states is None else states[start:end]
+                )
+                start = end
```

Comment:
Might want to add a check to see if the provided states are fully consumed by the model, otherwise it will be a hard to debug error.

Reply:
I think this already happens in `base_rnn.py`:

https://github.com/mergian/keras/blob/315bf5134aee6e69b516b136a192f37e25d5fa04/keras/layers/rnn/base_rnn.py#L928-L933